### ISSUE TITLE
RootPainter and RVE compatibility issue

### ISSUE THREAD

**Issue by eusunhan** [noble-research-institute/RhizoVisionExplorer#2]
Dear RVE team and many Rooties out there! 
Thank you very much for making this very useful software! This is about compatibility issue between two software: RootPainter and RVE.  

I am struggling to directly use the PNG files (segmentation output) from RootPainter software for analysis in RVE. It loads the PNG well, but the analysis can't be completed (stops around skeletonizing or the software crashes when I attempt to run it). BUT, When I convert the PNGs to JPEGs it works really great in RVE. (Or May be I am doing something wrong?)

As the two software can be used together by many scientists, it will be great to use the output from RootPainter directly in RVE. I attach both PNG and JPEG. 

Thank you.
BR, Eusun
![FVI_3_07_40-50_A_4000_01](https://user-images.githubusercontent.com/75310544/121176791-f9b9f100-c89f-11eb-8896-0d9fbab2c907.jpg)

![FVI_3_07_40-50_A_4000_01](https://user-images.githubusercontent.com/75310544/121176652-cd9e7000-c89f-11eb-9047-23f1c01ab4cb.png)

**Comment by eusunhan** [comment]
Dear Larry,
Thank you so much for detailed explanation. And it is great to know that there are ways around it. And yeah, I totally agree with you that it is not a bug of RVE. 

1. I will try out the removing alpha channel when using the PNG from RootPainter, which makes good sense.  
2. When converted to JPEG- I will use invert option. 

The RootPainter team might also consider tweaking the output transparency - and I hope both teams can talk about this issue together and make the two novel software more compatible soon. 

Thank you for the efforts and for great functionality of RVE. 
BR, Eusun

**Comment by Abe404** [comment]
Dear Larry and Eusun,

> I think it is typical for a segmented output to be black on white, with black encoding foreground and white as the background. [..] I would suggest you standardize the segmented output from RootPainter to be a true binary image with black roots on a white background.

I put 'binary segmentation' into google images and it appears to me that the opposite is true which also simplifies the analysis / data processing in RootPainter (for the libraries I use, working with existing datasets etc).

> I don't think use of transparency in image analysis is good practice, and we tried to be super clear in our documentation about what is expected. Transparencies are ambiguous because you need to let the software know if it should be interpreted as white, black, or what.

The RootPainter segmentation file format is designed to be loaded directly in the interface above the original photograph. I like having the data on the file system being the same as it is when loaded in the interface as this simplifies debugging/inspecting data with standard image viewers. That said, I appreciate that you might not want to support this type of transparent file as it adds complexity to RVE that is likely RootPainter specific.

> In the meantime, use of the JPG image does work but remember to use the 'Invert image' option.

Converting RootPainter PNGs to JPEG is not super hard but it might not be straight forward for users with large datasets who are less familiar with image processing. Converting to JPEG may also slightly degrade segmentation quality. This impact is probably usually negligible but is worth taking into account. I also believe that for binary segmentation, a PNG file will often have competitive file size.

I'd like to make it easier for users to directly process the output from RootPainter with RVE. I buy your arguments about transparency and appreciate it might not be straight forward for you to support RootPainter segmentations as they are now.

Given that you have already implemented the 'invert image' option, then I think a compromise is available that allows RootPainter to generate segmentations that can be directly processed by RVE. 

I could add a dropdown or radio button option when segmenting datasets in RootPainter, that allows a binary image to be generated with white foreground and black background. To enable compatibility with existing scripts and projects, I would still have the current option as the default (RP classic segmentation format with transparency) but could also mention in the interface that the second option (binary segmentation without alpha channel) enables integration with RVE,  (requiring the invert option  to be set in RVE). 

Improving integration should hopefully enable more convenient processing of a broader variety of root traits for RootPainter users and make it easier for RVE users to handle images with complex backgrounds.

I will also need to change the measurements and composite extraction functionality in RootPainter to handle these new segmentation files, but it shouldn't be too hard if we can go with this proposal.

@lmyork and @eusunhan would this solve the problem for now? And allow PNG output from RootPainter to be more conveniently processed with RVE? (without additional conversion steps in-between).

Kind regards,
Abraham

**Comment by lmyork** [comment]
Hi Abraham,

Thanks for your reply, providing clarity, and proposing an outcome.

You are absolutely correct that segmented images do seem to more commonly encode the foreground as white and background as black, but my main point was that they are binary, black and white images without transparency.

I had surmised that the transparency was likely for visualization purposes, which makes complete sense. I would assume, on the other hand, that transparency could be added internally from binary images as well, but that this might be complicated to implement initially. 

Turns out, in thinking about this more, that it may not actually be the transparency, really, but what the transparency is hiding, so we may wish to 'debug' a bit more? I think RVE is actually just ignoring the alpha channel. I am not totally sure, but I think the background of the RootPainter PNG is otherwise RGB(0,0,0), black. The blue-green color used in the RootPainter for the root is actually RGB(0,255,255). Our handling of color images was only meant to allow basic functions without full color-thresholding features, which we knew would be complicated and, indeed, lead to needs like RootPainter. So, unfortunately, we actually only use the red channel from color images. Thus, for the RootPainter PNG, the roots are red channel 0 and the background is also red channel 0. This may be the actual root cause. 

Therefore, using another color that includes red for the root segmented in RootPainter might work. Technically even RGB(1,255,255) should work, but I would recommend something like RGB(50, 255, 255) to make it a bit more different from 'black' in that channel. This is assuming I'm right that the background is somehow RGB(0,0,0) - or is it possible OpenCV internally drops the transparency and 'fills' in with black? This would still require invert option then in RVE.

On the other hand, if the hidden background could be changed to white RGB(255,255,255), then the foreground root could be kept at RGB(0,255,255) and RVE would essentially interpret this as a binary image with black roots on a white background, and the invert option would not be needed in this case. Therefore, @Abe404 is there a way for you to test this implementation and would it create any other issues for you? That is, if the background hidden by the alpha channel is black, can you simply make it white instead? Perhaps this is all that's needed...

Let me know if this makes sense, I know it's a bit complicated to put down to words.

Many thanks,

Larry

**Comment by Abe404** [comment]
Thanks for the swift and helpful response Larry,

> I would assume, on the other hand, that transparency could be added internally from binary images as well, but that this might be complicated to implement initially. [..]  if the hidden background could be changed to white RGB(255,255,255), then the foreground root could be kept at RGB(0,255,255) and RVE would essentially interpret this as a binary image with black roots on a white background, [..] Therefore, @Abe404 is there a way for you to test this implementation and would it create any other issues for you? 

You are correct and I think your proposals sound pragmatic. I have some concerns regarding compatibility and have a new proposal for an RVE specific export.

### Changing hidden background to white

There could be an issue where a new version of a RootPainter server would generate segmentations that a user with an old RootPainter client (painter) would try to process, causing bugs with the old client until the users upgrade. This would happen if going with your proposal of changing the hidden background to white RGB(255,255,255) as RootPainter currently assumes the blue channel is 0 in background regions when extracting root length or region properties.

I feel enabling backwards compatibility will make existing users more likely to upgrade to newer versions of RootPainter, as all their existing projects could still be opened and existing scripts will still work with RootPainter output segmentations old and new - unless they make a conscious choice to change the output format.

### Adjusting the Red channel

> RGB(1,255,255) should work, but I would recommend something like RGB(50, 255, 255)

This should work with RootPainter as it wouldn't cause issues for the existing length or region properties measurements. I'm not sure about the aesthetic yet. There is still a risk of breaking compatibility with any third party scripts and it still requires the user remembers to invert the image in RVE.

I think it's a trade-off. I may be over-estimating the importance of these changes for backwards compatibility.

### RootPainter to RVE export option

I could add another option to the extras menu which converted the segmentations generated from RootPainter to a format that is most compatible with RVE (in a new output folder specified by the user). That way I avoid any concerns regarding compatibility (with both old RootPainter versions and third party scripts written by other research groups) and make the conversation to an RVE ready format explicit. Perhaps having this explicit export step could give both teams a little more freedom and independence as it means the applications aren't too tightly coupled in ways they don't need to be (as long as the RVE specific export works).

I could even go with black foreground and white background to make the process easier for the user (they don't have to remember to set 'invert image' in RVE). These segmentations wouldn't need to have any compatibility with RootPainter or existing scripts and would be generating specifically for RVE. The downside to this is that it's an extra processing step to be done after the segmentation and an extra folder of files. But I expect it to be super quick, given that it's just a conversation from the RGBA image to a binary segmentation and the file sizes should be very small. I'm aware of some RootPainter deployments where the server component is managed by a separate administrator. This proposed RVE export functionality could work using only the client, simplifying the upgrade path for users who want to try this feature without coordinating with the admin maintaining the server.

Adding this option to the RootPainter extras menu could also be a way to highlight the option of integrating with RVE. The option can just be called 'Export for RhizoVisionExplorer' or 'convert segmentations for RhizoVisionExplorer'.

The RVE specific export functionality could be something that we revisit in the future to expand ways that RVE can take advantage of outputs from RootPainter. I would like to discuss this more at a future date - once RootPainter has some new features I have been planning.

Kind regards,
Abraham

**Comment by Abe404** [comment]
The other plus of the export/convert function is that it would be super quick and easy to implement.

**Comment by Abe404** [comment]
The new conversion utility seems to be working well and is very fast. I'm working on building the new cross-platform installers and then will share so @eusunhan can hopefully test the full 'pipeline' involving both software applications and make sure it works for his use case.

@lmyork I'm fine with keeping this here but let me know if you'd like us to move this discussion somewhere else as it's not an issue with RVE per se.

**Comment by lmyork** [comment]
Hi Abraham,

The client export utility would certainly be great. I leave it up to you whether to make reference to RhizoVision Explorer explicit, as users may want to use with WinRhizo or other software. On the other hand, I'm not sure PNG even works with WinRhizo (seem to remember it doesn't)? RVE can open commonly used image formats since it uses the OpenCV image read functions. You might consider "Export Binary Black-and-White PNGs" or "Convert to Binary Black-and-White PNGs" if you want to keep it general. Exporting to compressed TIFs might allow cross-functionality with WinRhizo. But, I do kind of prefer PNGs myself (for little reason).

In the future, we can add some basic color thresholding options on our side, I hope. Perhaps choosing a primary channel for one, but also color combinations to set a 'color space area' for thresholding. 

One question, I'm not clear if this export requires a batch processing of a folder to the RootPainter transparent PNGs, and then a second batch process to convert those? Is that the idea? Or can the BW PNG export be done directly after training (skipping the need to export the transparent PNGs first)? Either is OK, as long as we have clear instructions.

I'm happy to keep this conversation here. In the future, I will plan to include a page about RootPainter in our manual with instructions to do this export, as well.

Larry

**Comment by Abe404** [comment]
Hi Larry,

>The client export utility would certainly be great. I leave it up to you whether to make reference to RhizoVision Explorer explicit.

See inserted image:
<img width="547" alt="Screenshot 2021-06-13 at 17 31 32" src="https://user-images.githubusercontent.com/376295/121813703-8cc2a480-cc6d-11eb-81ff-12197fb0ae05.png">

The functionality is also available when a project is open.

installers coming (very) soon...

> In the future, we can add some basic color thresholding options on our side, I hope. Perhaps choosing a primary channel for one, but also color combinations to set a 'color space area' for thresholding.

Sounds like it could be useful, but you also can't support everything. Hopefully this new RootPainter conversion utility gives you one less input image use case to worry about. I also hope that maybe RootPainter could be useful to do the heavy lifting with the segmentation, for some of the use cases not yet covered by RVE. 

> One question, I'm not clear if this export requires a batch processing of a folder to the RootPainter transparent PNGs, and then a second batch process to convert those? Is that the idea? Or can the BW PNG export be done directly after training (skipping the need to export the transparent PNGs first)? Either is OK, as long as we have clear instructions.

It will require two processes. First segmentation and then conversation (outputting to a new folder). In my initial tests the conversion is very fast. Having this separate fast conversion process also means that research groups that have already segmented many images with RootPainter will have a quick way to get their data ready for RVE without having to re-segment (which requires server usage / file sync and takes longer as it means predicting all the roots again). I will discuss with Eusun regarding instructions / documentation. I think he might be working on a new video that could be a good place to showcase this new process. A html / markdown / pdf / colab notebook tutorial could also be considered.

> I'm happy to keep this conversation here. In the future, I will plan to include a page about RootPainter in our manual with instructions to do this export, as well.

Sounds great!

Abraham

**Comment by lmyork** [comment]
This is awesome, Abraham! Thanks for working to integrate the software, it really does seem like a 'dream team.'

I suggest Eusun can close this issue with our resolution that RootPainter now supports a conversion feature for generating binary BW images that can be used in RhizoVision Explorer. Let's stay in touch by e-mail and think about how to promote this pipeline.

**Comment by Abe404** [comment]
The release with the new installers are available here:

https://github.com/Abe404/root_painter/releases/tag/0.2.7

@eusunhan I suggest to test out the new version (downloads available from the above link) when you get chance in order to confirm that the issue has now been resolved for you, and then per Larry's suggestion to resolve this GitHub issue.

I have also created a discussion topic specifically for the new release (GitHub suggested I do so):
https://github.com/Abe404/root_painter/discussions/10

Although adding issues or email is also fine.

Thanks both,

Abraham

**Comment by asn5d** [comment]
Hi Abraham,

Thank you for the fix!

I will do a quick test for integration of the softwares.

Thanks,
Anand

**Comment by eusunhan** [comment]
Amazing!! Thanks to the creators of both software to make this work so quickly!

**Comment by Abe404** [comment]
> Hi Abraham,
> 
> Thank you for the fix!
> 
> I will do a quick test for integration of the softwares.
> 
> Thanks,
> Anand

Hi Anand, 

I'm just wondering how the testing went? Are the converted segmentations optimal for RVE?

Kind regards,
Abraham

**Comment by asn5d** [comment]
Hi Abraham,

The test was successful. I segmented the images that I trained earlier, converted to RVE optimal images and tested the images by loading them to RVE. I went without any problems, except that I had to change few lines of the RootPainter code to run on windows and change the number of workers to 0 (Line 221 at [trainer.py](https://github.com/Abe404/root_painter/blob/master/trainer/trainer.py)) to run on an older GPU (1080 gtx).

Thank you for making the fix! It is now easier to use the RootPainter-RVE workflow.

Thanks,
Anand

**Comment by Abe404** [comment]
Hi Anand,

I'm glad it works. If you have time to document, I'd be interested to know more details about what went wrong when you tried to run RootPainter on windows prior to your modifications. I assume the issue all issues were with the server component? 

Useful details would include:

1. Stack trace + error message
2. Expected behaviour
3. Actual behaviour
4. Your version of windows
5. Your python version number
6. Your pytorch version number
7. Your torchvision version number

No worries if you don't have time.

My initial plan was to only support linux for the server component (mainly because that's what I use for testing/development) but it seems multiple people are now running it on windows. If I can collect some details of the issues, then it's possible I can reproduce and then fix them next time I get access to a windows machine with a suitable GPU.

Kind regards,
Abraham
